File: lib/wModel/vocabulary.cpp
#include "vocabulary.h"

M::Vocabulary::Vocabulary(const W::Address p_address, QObject* parent):
    M::Model(p_address, parent),
    data(new W::Vocabulary())
{
    W::Handler* get = W::Handler::create(address + W::Address({u"get"}), this, &M::Vocabulary::_h_get);
    addHandler(get);
}

M::Vocabulary::~Vocabulary()
{
    delete data;
}

void M::Vocabulary::h_subscribe(const W::Event& ev)
{
    M::Model::h_subscribe(ev);
    
    h_get(ev);
}

void M::Vocabulary::h_get(const W::Event& ev)
{
    W::Vocabulary* vc = new W::Vocabulary;
    vc->insert(u"data", data->copy());
    
    response(vc, W::Address({u"get"}), ev);
}

void M::Vocabulary::insert(const W::String& key, const W::Object& value)
{
    if (registered) {
        W::Vocabulary* vc = new W::Vocabulary();
        W::Vocabulary* insert = new W::Vocabulary();
        W::Vector* erase = new W::Vector();
        
        if (data->has(key)) {
            erase->push(key);
        }
        data->insert(key, value);
        insert->insert(key, value);
        
        vc->insert(u"insert", insert);
        vc->insert(u"erase", erase);
        
        broadcast(vc, W::Address{u"change"});
    } else {
        data->insert(key, value);
    }
}

void M::Vocabulary::insert(const W::String& key, W::Object* value)
{
    if (registered) {
        W::Vocabulary* vc = new W::Vocabulary();
        W::Vocabulary* insert = new W::Vocabulary();
        W::Vector* erase = new W::Vector();
        
        if (data->has(key)) {
            erase->push(key);
        }
        data->insert(key, value);
        insert->insert(key, value->copy());
        
        vc->insert(u"insert", insert);
        vc->insert(u"erase", erase);
        
        broadcast(vc, W::Address{u"change"});
    } else {
        data->insert(key, value);
    }
}

void M::Vocabulary::erase(const W::String& key)
{
    data->erase(key);
    if (registered) {
        W::Vocabulary* vc = new W::Vocabulary();
        W::Vocabulary* insert = new W::Vocabulary();
        W::Vector* erase = new W::Vector();
            
        erase->push(key);
            
        vc->insert(u"insert", insert);
        vc->insert(u"erase", erase);
        
        broadcast(vc, W::Address{u"change"});
    }
}

void M::Vocabulary::clear()
{
    data->clear();
    
    if (registered) {
        broadcast(new W::Vocabulary(), W::Address{u"clear"});
    }
}

M::Model::ModelType M::Vocabulary::getType() const
{
    return type;
}

void M::Vocabulary::set(const W::Object& value)
{
    delete data;
    data = static_cast<W::Vocabulary*>(value.copy());
    
    W::Vocabulary* vc = new W::Vocabulary;
    vc->insert(u"data", data->copy());
    
    broadcast(vc, W::Address({u"get"}));
}

void M::Vocabulary::set(W::Object* value)
{
    delete data;
    data = static_cast<W::Vocabulary*>(value);
    
    W::Vocabulary* vc = new W::Vocabulary;
    vc->insert(u"data", data->copy());
    
    broadcast(vc, W::Address({u"get"}));
}

File: libjs/wModel/vocabulary.js
"use strict";

var Model = require("./model");
var Vector = require("../wType/vector");
var Vocabulary = require("../wType/vocabulary");
var Object = require("../wType/object")
var String = require("../wType/string");

var ModelVocabulary = Model.inherit({
    "className": "Vocabulary",
    "constructor": function(address) {
        Model.fn.constructor.call(this, address);
        
        this._data = new Vocabulary();
        
        this.addHandler("get");
    },
    "destructor": function() {
        this._data.destructor();
        
        Model.fn.destructor.call(this);
    },
    "clear": function() {
        this._data.clear();
        
        if (this._regestered) {
            this.broadcast(new Vocabulary(), "clear");
        }
    },
    "erase": function(key) {
        this._data.erase(key);
        
        if (this._registered) {
            var vc = new Vocabulary();
            var insert = new Vocabulary();
            var erase = new Vector();
            
            erase.push(new String(key));
            vc.insert("insert", insert);
            vc.insert("erase", erase);
            
            this.broadcast(vc, "change");
        }
    },
    "_h_subscribe": function(ev) {
        Model.fn._h_subscribe.call(this, ev);
        
        this._h_get(ev);
    },
    "_h_get": function(ev) {
        var vc = new Vocabulary();
        
        vc.insert("data", this._data.clone());
        this.response(vc, "get", ev);
    },
    "insert": function(key, value) {
        if (this._registered) {
            var vc = new Vocabulary();
            var insert = new Vocabulary();
            var erase = new Vector();
            
            if (this._data.has(key)) {
                erase.push(new String(key));
            }
            this._data.insert(key, value);
            insert.insert(key, value.clone());
            
            vc.insert("insert", insert);
            vc.insert("erase", erase);
            
            this.broadcast(vc, "change");
            
        } else {
            this._data.insert(key, value);
        }
    }
});

module.exports = ModelVocabulary;
